src/nvim/statusline.c

```diff
@@ -1640,18 +1641,11 @@ int build_stl_str_hl(win_T *wp, char *out, size_t outlen, char *fmt, OptIndex op
 
       char *p = NULL;
       if (fold) {
-        schar_T fold_buf[10];
-        size_t n = fill_foldcolumn(NULL, wp, stcp->foldinfo,
-                                   (linenr_T)get_vim_var_nr(VV_LNUM), 0, fold_buf);
+        size_t n = fill_foldcolumn(wp, stcp->foldinfo, (linenr_T)get_vim_var_nr(VV_LNUM),
+                                   fdc, NULL, out_p);
         stl_items[curitem].minwid = -((stcp->use_cul ? HLF_CLF : HLF_FC) + 1);
-        size_t buflen = 0;
-        // TODO(bfredl): this is very backwards. we must support schar_T
-        // being used directly in 'statuscol'
```

Comment:
Is this realistic? I guess that would require immediately converting to `schar_T` and placing on the screen while parsing inside `build_stl_str_hl()`. Seems like an unlikely refactor. Wasn't `fill_foldcolumn()` prematurely made to put characters on the screen directly? Since 'statuscolumn' is a consumer as well, filling a `char` buffer and letting `draw_col_buf()` convert to `schar_T` for the default 'foldcolumn' seems more appropriate. Made a PR with a concrete change instead of discussing about it but feel free to prove me wrong.

Reply:
> Wasn't fill_foldcolumn() prematurely made to put characters on the screen directly? 

There's absolutely nothing "premature" with using screen chars and screen data structures to render text to be put on the screen. In addition `wp->w_p_fcs_chars` is going to be changed to use `schar_T` directly (just like signs have), so `fill_foldcolumn()` is going to need to deal with `schar_T` values regardless.

> I guess that would require immediately converting to schar_T and placing on the screen while parsing inside build_stl_str_hl(). Seems like an unlikely refactor considering it's complexity.

This sounds weirdly and very specifically defeatist, given the even more complex refactors needed to get win_line(), grid.c etc to the state where they are today. But regardless, it is not something that has to be done immediately. The point is extra complexity of stl implementation being an outlier has to be contained within it, is should not disrupt the flow of the code all around it, like drawline.c . After all the _primary purpose_ of statuslines and -columns is to be rendered on the screen, `nvim_eval_statusline` is useful to have but it is a secondary purpose and it is fine if that's what's going to need an extra conversion.

This change just seems like code golfing back into a previous local minimum instead of driving refactor work forward.

Reply:
> There's absolutely nothing "premature" with using screen chars and screen data structures to render text to be put on the screen.

Premature in the sense that there are other callers of the function that do not do that. But OK got it.

Reply:
> The point is extra complexity of stl implementation being an outlier has to be contained within it, is should not disrupt the flow of the code all around it, like drawline.c .

I mean the current implementation only introduces extra complexity in drawline.c

Reply:
> I mean I would say with the current implementation, disrupting the code and introducing extra complexity in drawline.c is exactly whats happening. Previous commit of this PR only reduced that IMO.

Well I have to disagree with that. The simple way to put text on the screen is to just assign it to `linebuf_char` and add metadata to its siblings attr/vcol as needed at the same time. moving the treatment of `linebuf_vcol` foldcolumn values to `draw_col_buf` would be a crude hack which is not less disruptive just because it occupies fewer lines of code.  Not to mention, the fact that these values (i e click behavior) get lost in 'statuscol' is still something to consider a bug (whose fix would follow easily if stl code became more similar to the builtin column code and used screen buffers, _not_ the other way around).

Reply:
> Well I have to disagree with that. The simple way to put text on the screen is to just assign it to linebuf_char and add metadata to its siblings attr/vcol as needed at the same time. moving the treatment of linebuf_vcol foldcolumn values to draw_col_buf would be a crude hack which is not less disruptive just because it occupies fewer lines of code. 

Yeah I can agree to that POV. Though if `draw_col_buf` is there to stay, I don't see the harm in using it.

> Not to mention, the fact that these values (i e click behavior) get lost in 'statuscol' is still something to consider a bug (whose fix would follow easily if stl code became more similar to the builtin column code and used screen buffers, not the other way around).

I'm not sure that it should be considered a bug, moreso an inconvenience perhaps. 'statuscolumn' can assign its own click callback handlers (`mouse_check_grid()` is bypassed entirely).

Reply:
>  Though if draw_col_buf is there to stay, I don't see the harm in using it.

It might be or it might not be (see the comment above it)